pkg/volume/azure_dd/azure_dd.go

```diff
@@ -189,6 +196,26 @@ func (plugin *azureDataDiskPlugin) NewUnmounter(volName string, podUID types.UID
 	}, nil
 }
 
+func (plugin *azureDataDiskPlugin) RequiresFSResize() bool {
+	return true
+}
+
+func (plugin *azureDataDiskPlugin) ExpandVolumeDevice(
+	spec *volume.Spec,
+	newSize resource.Quantity,
+	oldSize resource.Quantity) (resource.Quantity, error) {
+	if spec.PersistentVolume == nil || spec.PersistentVolume.Spec.AzureDisk == nil {
+		return oldSize, fmt.Errorf("invalid PV spec")
+	}
+
+	diskController, err := getDiskController(plugin.host)
+	if err != nil {
+		return oldSize, err
+	}
+
+	return diskController.ResizeDisk(spec.PersistentVolume.Spec.AzureDisk.DiskName, oldSize, newSize)
```

Comment:
nit: Add a info log here with diskname, oldSize and newSize

Reply:
How fast is resize operation on Azure btw? If it is slow and controller crashes while resize is in progress, I am wondering if there is a way to query the volume and find out if resize is in progress with cloudprovider and return early from here rather than calling resize again.

I know GCE and AWS does not implement this mechanism currently but I am going to change it for AWS at least because resize is pretty slow there.

Reply:
@gnufied azure disk resize operation could be completed in a few seconds, while Azure does not support resize when that disk is attached to a running VM, GCE and AWS supports that, right?

Reply:
good point, shouldn't `ResizeDisk` call then check if volume is attached somewhere before allowing volume resize operation? Expand controller does not have any kind of fencing to prevent volumes from being resused when mounted on nodes.

Reply:
@gnufied if that disk is already attached to a running VM, the API call will fail with error, eventually `ResizeDisk` will return error, that's expected case.

Reply:
I don't know much about Azure tbh, but I have found that it is less costly to make a `GET` request and return early than make a `POST` request and handle error and return (cloudproviders seem to have different quotas for mutable and immutable API requests).

And as far as I can tell, you are already fetching disk state from cloudprovider - https://github.com/kubernetes/kubernetes/pull/64386/files#diff-db9a5ad5d2cc7740ca2b73ad9d904fa2R144, so if we can ascertain that volume is attached, we should return error before even attempting to call resize.

This is not a blocker though, I know very little about Azure - so take my comment with a pinch of salt. :-)

Reply:
@gnufied thanks for the check, while according to https://docs.microsoft.com/en-us/rest/api/compute/disks/get, the returned value `DiskProperties.ProvisioningState` is not sufficient to tell whether disk is in `Attached` state or not, we need another `GET` request to get `diskState`. So for this PR, I think it's ok to do resize directly.